File: backend/app/core/permissions.py
import enum

from app.db.models.user import UserRole


class Permission(enum.StrEnum):
    DASHBOARD_READ = "dashboard.read"
    ROOM_READ = "room.read"
    ROOM_MANAGE = "room.manage"
    CANDIDATE_READ = "candidate.read"
    CANDIDATE_MANAGE = "candidate.manage"
    SESSION_READ = "session.read"
    SESSION_MANAGE = "session.manage"
    SESSION_MONITOR = "session.monitor"
    MEDIA_READ = "media.read"
    MEDIA_UPLOAD = "media.upload"
    TRACKING_READ = "tracking.read"
    EVENT_READ = "event.read"
    EVENT_CREATE_MANUAL = "event.create_manual"
    EVENT_REVIEW = "event.review"
    EVIDENCE_READ = "evidence.read"
    EVIDENCE_MANAGE = "evidence.manage"
    APPEAL_READ = "appeal.read"
    APPEAL_MANAGE = "appeal.manage"
    REPORT_READ = "report.read"
    REPORT_EXPORT = "report.export"
    USER_MANAGE = "user.manage"
    AUDIT_READ = "audit.read"
    AUDIT_READ_RELEVANT = "audit.read_relevant"
    SYSTEM_MANAGE = "system.manage"
    DIAGNOSTICS_READ = "diagnostics.read"


ALL_PERMISSIONS = frozenset(Permission)

ROLE_PERMISSIONS: dict[UserRole, frozenset[Permission]] = {
    UserRole.SUPERVISOR: frozenset(
        {
            Permission.DASHBOARD_READ,
            Permission.ROOM_READ,
            Permission.CANDIDATE_READ,
            Permission.CANDIDATE_MANAGE,
            Permission.SESSION_READ,
            Permission.SESSION_MANAGE,
            Permission.SESSION_MONITOR,
            Permission.MEDIA_READ,
            Permission.MEDIA_UPLOAD,
            Permission.TRACKING_READ,
            Permission.EVENT_READ,
            Permission.EVENT_CREATE_MANUAL,
            Permission.EVIDENCE_READ,
            Permission.REPORT_READ,
        }
    ),
    UserRole.REVIEWER: frozenset(
        {
            Permission.DASHBOARD_READ,
            Permission.ROOM_READ,
            Permission.CANDIDATE_READ,
            Permission.SESSION_READ,
            Permission.TRACKING_READ,
            Permission.EVENT_READ,
            Permission.EVENT_REVIEW,
            Permission.EVIDENCE_READ,
            Permission.EVIDENCE_MANAGE,
            Permission.APPEAL_READ,
            Permission.APPEAL_MANAGE,
            Permission.REPORT_READ,
            Permission.AUDIT_READ_RELEVANT,
        }
    ),
    UserRole.ADMIN: ALL_PERMISSIONS,
}


def permissions_for_role(role: str | UserRole) -> frozenset[Permission]:
    try:
        resolved_role = UserRole(role)
    except ValueError:
        return frozenset()
    return ROLE_PERMISSIONS[resolved_role]


def has_permission(role: str | UserRole, permission: Permission) -> bool:
    return permission in permissions_for_role(role)

File: frontend/src/features/auth/permissions.ts
import { useCallback, useMemo } from 'react'

import { useAuth } from './AuthProvider'
import type { UserRole } from './types'

export const permissions = {
  dashboardRead: 'dashboard.read',
  roomRead: 'room.read',
  roomManage: 'room.manage',
  candidateRead: 'candidate.read',
  candidateManage: 'candidate.manage',
  sessionRead: 'session.read',
  sessionManage: 'session.manage',
  sessionMonitor: 'session.monitor',
  mediaRead: 'media.read',
  mediaUpload: 'media.upload',
  trackingRead: 'tracking.read',
  eventRead: 'event.read',
  eventCreateManual: 'event.create_manual',
  eventReview: 'event.review',
  evidenceRead: 'evidence.read',
  evidenceManage: 'evidence.manage',
  appealRead: 'appeal.read',
  appealManage: 'appeal.manage',
  reportRead: 'report.read',
  reportExport: 'report.export',
  userManage: 'user.manage',
  auditRead: 'audit.read',
  auditReadRelevant: 'audit.read_relevant',
  systemManage: 'system.manage',
  diagnosticsRead: 'diagnostics.read',
} as const

export type Permission = typeof permissions[keyof typeof permissions]

export const allPermissions = Object.freeze(Object.values(permissions)) as readonly Permission[]

export const rolePermissions: Readonly<Record<UserRole, ReadonlySet<Permission>>> = {
  SUPERVISOR: new Set([
    permissions.dashboardRead,
    permissions.roomRead,
    permissions.candidateRead,
    permissions.candidateManage,
    permissions.sessionRead,
    permissions.sessionManage,
    permissions.sessionMonitor,
    permissions.mediaRead,
    permissions.mediaUpload,
    permissions.trackingRead,
    permissions.eventRead,
    permissions.eventCreateManual,
    permissions.evidenceRead,
    permissions.reportRead,
  ]),
  REVIEWER: new Set([
    permissions.dashboardRead,
    permissions.roomRead,
    permissions.candidateRead,
    permissions.sessionRead,
    permissions.trackingRead,
    permissions.eventRead,
    permissions.eventReview,
    permissions.evidenceRead,
    permissions.evidenceManage,
    permissions.appealRead,
    permissions.appealManage,
    permissions.reportRead,
    permissions.auditReadRelevant,
  ]),
  ADMIN: new Set(allPermissions),
}

export function hasPermission(role: UserRole | null | undefined, permission: Permission): boolean {
  return role ? rolePermissions[role].has(permission) : false
}

export function usePermissions() {
  const { user } = useAuth()
  const granted = useMemo(
    () => user ? rolePermissions[user.role] : new Set<Permission>(),
    [user],
  )
  const can = useCallback((permission: Permission) => granted.has(permission), [granted])
  return { can, permissions: granted }
}
